python/cuda_parallel/cuda/parallel/experimental/algorithms/reduce.py

```diff
@@ -119,9 +131,28 @@ def __del__(self):
         bindings.cccl_device_reduce_cleanup(ctypes.byref(self.build_result))
 
 
+def make_cache_key(
+    d_in: DeviceArrayLike | IteratorBase,
+    d_out: DeviceArrayLike,
+    op: Callable,
+    h_init: np.ndarray,
+):
+    d_in_key = d_in if isinstance(d_in, IteratorBase) else cai.get_dtype(d_in)
+    d_out_key = d_out if isinstance(d_out, IteratorBase) else cai.get_dtype(d_out)
+    op_key = (op.__code__.co_code, op.__code__.co_consts, op.__closure__)
+    h_init_key = h_init.dtype
+    return (d_in_key, d_out_key, op_key, h_init_key)
```

Comment:
important: there's an implicit state affecting the key. As written, I'd get the same reducer object for devices of different architectures:

```python
cudaSetDevice(0)
reducer_1 = reduce_into(d_in, d_out, ...)
cudaSetDevice(1)
reducer_2 = reduce_into(d_in, d_out, ...)
```

Let's incorporate `cc_major, cc_minor = cuda.get_current_device().compute_capability` somewhere in the key to address that.

Reply:
Good call. I did that in https://github.com/NVIDIA/cccl/pull/3216/commits/56f2c61f774e965df75b55b9c3028711c5983a53.